### PR TITLE
Selection Pane Addition

### PR DESCRIPTION
**Pull request by D-moe**
Add support for selecting portions of runs, selection pane for multiple runs added (just the layout not the feature).
I forked this from the master branch which will not compile for java until Frelicia's branch merge fixes this problem so this error can be ignored.

### CODE REVIEW

File: src/main/webapp/2Dscript.js [hunk 1]
```diff
@@ -80,7 +151,279 @@ function addCenterControl(controlDiv) {
...
+      }
+
+    });
+    checkBoxEntry.appendChild(input);
+    currentRow.appendChild(checkBoxEntry);
+    table.appendChild(currentRow);
+    console.log(currentRow);
+    even = !even;
+
+  }
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:230]: This loop is huge; please refactor into smaller functions.

File: src/main/webapp/2Dscript.js [hunk 2]
```diff
@@ -60,12 +106,37 @@ function formatPoseData() {
...
+  const selectionUI = document.createElement('div');
+  selectionUI.className = 'selectionUI';
+  selectionUI.title = 'Select the pose runs to view/render in 3D.';
+  sideControlDiv.appendChild(selectionUI);
+
+  // Set CSS for the selectin pane interior.
+  const selectionText = document.createElement('div');
+  selectionText.className = 'selectionText';
+  selectionText.appendChild(innerContent);
+  // Need to call a function to create a table with checkboxes here, likely write some generic function to do it with any JSON file possible. Also add checkbox as column and an eyeball icon with event listeners.
```
[the remark this answers is not in the document]
Comment (D-moe) [reply src/main/webapp/2Dscript.js:129]: Yeah still need to resolve some of the linter errors, this is one of them

File: src/main/webapp/2Dscript.js [hunk 3]
```diff
@@ -80,7 +151,279 @@ function addCenterControl(controlDiv) {
...
+    table.appendChild(currentRow);
+    console.log(currentRow);
+    even = !even;
+
+  }
+  return table;
+
+}
+
+function plotLine(dataEntries) {
```
[the remark this answers is not in the document]
Comment (D-moe) [reply src/main/webapp/2Dscript.js:235]: ditto to linter error

File: src/main/webapp/2Dscript.js [hunk 4]
```diff
@@ -80,7 +151,279 @@ function addCenterControl(controlDiv) {
...
+        console.log("The value of dataentries is " + dataEntries);
+        // if the value is not found in the cache then fetch it.
+        if (dataEntries === undefined) {
+          console.log("Data Entries was false");
+          console.log("The id is:" + event.target.id);
+          fetch("/getrun?id=" + event.target.id + "&dataType=pose").then(response => response.json())
+            .then(data => dataEntries = data).then(() => {
+              datas[event.target.id] = dataEntries;
+              const toGraph = plotLine(dataEntries);
+              maps[event.target.id] = toGraph;
```
[the remark this answers is not in the document]
Comment (D-moe) [reply src/main/webapp/2Dscript.js:202]: What's the rationale for this change, is it more efficient or less error prone?

File: src/main/webapp/2Dscript.js [hunk 5]
```diff
@@ -80,7 +151,279 @@ function addCenterControl(controlDiv) {
 
   // Setup the click event listeners: simply set the map to Chicago.
   controlUI.addEventListener('click', function() {
-    window.location.href = '/3DVisual.html?id=' +
-    runId + '&dataType=' + dataType;
+    window.location.href = '/3DVisual.html?id=' + id + '&dataType=' + type;
   });
 }
+
+function createSideTable(json) {
```
[the remark this answers is not in the document]
Comment (D-moe) [reply src/main/webapp/2Dscript.js:158]: ditto linter

File: src/main/webapp/2Dscript.js [hunk 6]
```diff
@@ -1,56 +1,102 @@
...
-  // Intitalizes the map.
+  // Occurs when google map api calls all promises, which is getting this
+  // function called somehow, according to stack trace.
+  if (pose === undefined) {
+    return;
+  }
+
+  console.log('initMap called');
+  const poseCoordinates = [];
+  for (let i = 0; i < pose.length; i++) {
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:56]: Can you use `for (const point of pose)`?

https://google.github.io/styleguide/jsguide.html#features-for-loops

File: src/main/webapp/2Dscript.js [hunk 7]
```diff
@@ -1,56 +1,102 @@
...
+// Define bounding box around the different data sets to quickly determined
+// whether they should be included in the calculation or not
 
+/**
+ * A Point object with a lat and lng.
+ * @typedef {Object<number, number>} Point
+ * @property {number} lat The lat of the point.
+ * @property {number} lng The lng of the point.
+ */
+let flightPath;
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:20]: What is `flightPath`? The name is surprising to me.

File: src/main/webapp/2Dscript.js [hunk 8]
```diff
@@ -80,7 +151,279 @@ function addCenterControl(controlDiv) {
...
+    strokeOpacity: 1.0,
+    strokeWeight: 2
+  });
+  flightPath.setMap(null);
+  console.log("The value of map is " + map);
+  return currentLineGraph;
+}
+
+// When finished drawing the box, instead of leaving the widget, create markers at the current lat and intersection with the line.
+$(function () {
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:255]: This is another huge and deeply nested function, please refactor. I am not familiar with the `$` syntax, what is it doing?

File: src/main/webapp/2Dscript.js (hunk 1 above)
Comment (D-moe) [reply src/main/webapp/2Dscript.js]: Yeah was working on that wanted to get it pushed though, so people could see what was going on.

File: src/main/webapp/2Dscript.js (hunk 8 above)
Comment (D-moe) [reply src/main/webapp/2Dscript.js]: I used Jquery in this code to simplify the many nested event handlers $( gives a global selector in jquery

File: src/main/webapp/2Dscript.js [hunk 9]
```diff
@@ -1,56 +1,102 @@
 // Gobal variables.
```
[the remark this answers is not in the document]
Comment (D-moe) [reply src/main/webapp/2Dscript.js]: My bad, should have been gobble like a turkey

[the remark this answers is not in the document]
Comment (D-moe) [reply src/main/webapp/2Dscript.js]: will correct

File: src/main/webapp/2Dscript.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (zjwoody) [reply src/main/webapp/2Dscript.js]: The way of setting a property using [] does not interact with the Map data structure. It uses the feature of the generic object. The value is not stored in the Map for queries. The operations like has/delete on the data will fail.
See details:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

[the remark this answers is not in the document]
Comment (D-moe) [reply src/main/webapp/2Dscript.js]: I guess I don't understand why we should use a map instead of an object in this case? As this is a simple string key pairing, it seems like these limitations are acceptable. Also, its possible we may want to convert this map object into a json string for passing through sessionstorage etc, a limitation that a map doesn't allow I believe? https://www.geeksforgeeks.org/map-vs-object-in-javascript/ (for reference on the JSON conversion issue)

[the remark this answers is not in the document]
Comment (D-moe) [reply src/main/webapp/2Dscript.js]: I guess perhaps the naming convention was confusing as I titled it maps as it stores the lines that are graphed not because it is a dataMap.

File: src/main/webapp/2Dscript.js [hunk 10]
```diff
@@ -1,24 +1,43 @@
...
+ * from the requisite proto in the JSON format.
+ * Position is given in WGS84 latitude, longitude, and altitude.
+ * @typedef {Object<number, number, number, number,
+ * number, number, number>} PoseData
+ * @property {number} gpsTimestamp The time in seconds.
+ * @property {number} lat Degrees in [-90, 90]
+ * @property {number} lng Degrees in [-180, 180]
+ * @property {number} alt Meters above WGS84
+ * @property {number} rollDeg Degrees in [0,360]
+ * @property {number} pitchDeg Degrees in [0,360]
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:24]: I think pitch is actually in [-90, 90].

File: src/main/webapp/2Dscript.js [hunk 11]
```diff
@@ -80,7 +169,361 @@ function addCenterControl(controlDiv) {
...
+ * @param {number} valLat
+ * @param {number} valLng
+ * @return {boolean}
+ */
+function withinBound(minLat, maxLat, minLng, maxLng, valLat, valLng) {
+  if ((valLat >= minLat && valLat <= maxLat) &&
+    (valLng >= minLng && valLng <= maxLng)) {
+    return true;
+  }
+  return false;
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:528]: This can be simplified:
`return lat >= minLat && lat <= maxLat && lng >= minLng && lng <= maxLng;`

File: src/main/webapp/2Dscript.js [hunk 12]
```diff
@@ -1,56 +1,122 @@
...
+const dataCache = {};
 let pose;
-let runId;
-let dataType;
+let data;
+let id;
+let type;
+let currentLat;
+let currentLong;
+let initialPoseData;
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:11]: It's getting pretty difficult to follow the data flow with all the global variables. Can you make these into local variables passed around explicitly as much as possible?

File: src/main/webapp/2Dscript.js [hunk 13]
```diff
@@ -80,7 +169,361 @@ function addCenterControl(controlDiv) {
...
+
+
+/**
+ * Generate a polyline from the given data points and return it.
+ * @param {Array<Point>} dataEntries
+ * @return {google.maps.Polyline}
+ */
+function plotLine(dataEntries) {
+  const currentLine = [];
+  for (let i = 0; i < dataEntries.length; i++) {
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:337]: Can you use `for (point of dataEntries)`?

File: src/main/webapp/3Dscript.js [hunk 14]
```diff
@@ -208,3 +191,33 @@ function animate() {
...
+  const queryString = window.location.search;
+  const urlParams = new URLSearchParams(queryString);
+  const id = urlParams.get('id');
+  const type = urlParams.get('dataType');
+  // Only need to refetch the data if it is not contained in local storage, in
+  // general it should be.
+  const isStored = urlParams.get('stored');
+  const subSectionNumber = urlParams.get('subSection');
+  console.log(id + type);
+  if (!isStored) {
```
Comment (jmchristen) [inline src/main/webapp/3Dscript.js:209]: Prefer to test the positive condition first:
```
if (isStored) {
  ...
} else {
  ...
}
```
That way the reader doesn't have to do double negation.

https://drive.google.com/file/d/1-g637f-xAyGWbcILf5QbLRiTmvsDA3P9/view?usp=sharing

File: src/main/webapp/2Dscript.js (hunk 10 above)
Comment (D-moe) [reply src/main/webapp/2Dscript.js:24]: Copied this from the master branch proto comment, so would that mean it was wrong there too?

File: src/main/webapp/2Dscript.js (hunk 12 above)
Comment (D-moe) [reply src/main/webapp/2Dscript.js:11]: I'll work on moving some of these into functions. I know I was having a bit of a challenge figuring out the proper scope for variables passed to event handlers, any tips for that?

File: src/main/webapp/2Dscript.js (hunk 10 above)
Comment (jmchristen) [reply src/main/webapp/2Dscript.js:24]: Probably so.

File: src/main/webapp/2Dscript.js (hunk 12 above)
Comment (jmchristen) [reply src/main/webapp/2Dscript.js:11]: I'm not sure what you mean. Are you talking about something like:
```
onclick(event) {
  map.setSomething(event.something) // map is currently global
}
```
and wondering what the scope of `map` should be? Or something else?

Comment (D-moe) [reply src/main/webapp/2Dscript.js:11]: Yeah so it gets confusing to me if something like this occurs
function doSomething(){
let a;
let b;
b.addEventListener('click', function(){
// reference the a which is the scope of doSomething
console.log(a);
});
vs if you factor out this lambda into a function, passing that value to the function without using global scope. I tried using bind, but I think I was misunderstanding how to use 'this' in this case?

Comment (jmchristen) [reply src/main/webapp/2Dscript.js:11]: How about:
```
function doSomething() {
  let a;
  let b;
  b.addEventListener('click', function() { handleClick(a) });
}
function handleClick(a) {
  console.log(a);
}
```
There is still a lambda, but it's very short even if `handleClick` is long.

Comment (D-moe) [reply src/main/webapp/2Dscript.js:11]: for (int i =0 ; i <5; i++){
a.push(i)

Comment (D-moe) [reply src/main/webapp/2Dscript.js:11]: b[i].addEventListener

File: src/main/webapp/2Dscript.js [hunk 15]
```diff
@@ -80,7 +180,410 @@ function addCenterControl(controlDiv) {
...
+        position: latLng,
+      });
+      infoWindow.setMap(map);
+    },
+    );
+
+    subPath.setMap(map);
+
+    markerBottom = genMarker(discoveredMinLat, discoveredMinLatPair);
+    markerTop = genMarker(discoveredMaxLngPair, discoveredMaxLng);
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:462]: It seems odd to reassign `markerTop` and `markerBottom` on each iteration of the loop. What's the purpose?

File: src/main/webapp/2Dscript.js [hunk 16]
```diff
@@ -80,7 +180,410 @@ function addCenterControl(controlDiv) {
...
+  console.log(pose);
+  const poseLength = pose.length;
+  // Lat can be from [-90,90].
+  let discoveredMinLat = 91;
+  let discoveredMinLatPair = 181;
+  // Lng can be from [-180,180].
+  let discoveredMaxLng = -181;
+  let discoveredMaxLngPair = -91;
+  subLine = [];
+  for (let i = 0; i < poseLength; i++) {
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:550]: `for (point of pose)`

File: src/main/webapp/2Dscript.js [hunk 17]
```diff
@@ -80,7 +180,410 @@ function addCenterControl(controlDiv) {
...
+  $('.widget').remove();
+
+  const mapRuns = Object.entries(runs);
+  // Clear pop up window.
+  if (infoWindow !== undefined) {
+    infoWindow.setMap(null);
+  }
+  // Clear other paths.
+
+  for ([id, currentRun] of mapRuns) {
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:426]: Can the loop body be factored out into a function?

File: src/main/webapp/2Dscript.js [hunk 18]
```diff
@@ -80,7 +180,410 @@ function addCenterControl(controlDiv) {
...
+      console.log('Polyline clicked at lat: ' + latLng.lat() +
+        ' lng: ' + latLng.lng());
+      infoWindow = new google.maps.InfoWindow({
+        content: '<a href=/home.html?id=' + id + '&dataType=' +
+          type + '&subSection=' + subSectionNumber +
+          '&stored=true' + '> View in 3D </a>',
+        position: latLng,
+      });
+      infoWindow.setMap(map);
+    },
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:456]: Can the listener be factored out into a function?

File: src/main/webapp/2Dscript.js [hunk 19]
```diff
@@ -80,7 +180,410 @@ function addCenterControl(controlDiv) {
...
+  return polyLine;
+}
+
+/**
+ * Compute the subsection of a line contained in a bounding rectangle.
+ * @param {Point} pose
+ * @param {number} currentLat
+ * @param {number} priorLat
+ * @param {number} currentLng
+ * @param {number} priorLng
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:533]: I think `pose` is actually an Array? And current/prior is surprising terminology for a bounding rectangle; I would expect something like min/max.

File: src/main/webapp/2Dscript.js (hunk 15 above)
Comment (D-moe) [reply src/main/webapp/2Dscript.js:462]: In this case it doesn't really make sense to have the loop, this is a remnant for trying to add an additional feature, I'll remove.

File: src/main/webapp/2Dscript.js (hunk 19 above)
Comment (D-moe) [reply src/main/webapp/2Dscript.js:533]: Yeah should be Array<Point> good catch. The reasoning for this was the first click gives the priorLat and priorLng and the next click gives the currentLng and currentLng that you clicked to. That's why I used that terminology as I was thinking in terms of when the events happen not the bounds of the rectangle itself.

File: src/main/webapp/2Dscript.js [hunk 20]
```diff
@@ -80,7 +175,418 @@ function addCenterControl(controlDiv) {
...
+/**
+ * Generate a polyline from the given data points and return it.
+ * @param {Array<Point>} dataEntries
+ * @return {google.maps.Polyline}
+ */
+function plotLine(dataEntries) {
+  const currentLine = [];
+  for (point of dataEntries) {
+    currentLine.push({lat: point.lat, lng: point.lng});
+  }
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:343]: I think you can do
`const currentLine = [...dataEntries];`
or
`const currentLine = dataEntries.slice();`
instead.

File: src/main/webapp/2Dscript.js [hunk 21]
```diff
@@ -80,7 +175,418 @@ function addCenterControl(controlDiv) {
...
+  // Clear pop up window.
+  if (infoWindow !== undefined) {
+    infoWindow.setMap(null);
+  }
+  // Clear other paths.
+  // line
+  // minLatPoint
+  // lat , lng
+  // minLngPoint
+  // lat, lng
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:422]: Extraneous comment.

File: src/main/webapp/2Dscript.js [hunk 22]
```diff
@@ -80,7 +175,418 @@ function addCenterControl(controlDiv) {
...
+ * Create a polyline from the given points with specified parameters.
+ * @param {Point} linePoints
+ * @param {string} color
+ * @param {number} opacity
+ * @param {number} weight
+ * @param {number} index
+ * @return {google.maps.Polyline}
+ */
+function getPolyLine(linePoints, color, opacity, weight, index = 1) {
+  const polyLine = new google.maps.Polyline({
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:498]: Follow the API's lead on capitalization: "polyline" is all one word, so the L shouldn't be capitalized.

File: src/main/webapp/2Dscript.js [hunk 23]
```diff
@@ -80,7 +175,418 @@ function addCenterControl(controlDiv) {
...
+/**
+ * Compute the subsection of a line contained in a bounding rectangle.
+ * @param {Array<Point>} pose
+ * @param {number} currentLat
+ * @param {number} priorLat
+ * @param {number} currentLng
+ * @param {number} priorLng
+ * @return {subSectionData}
+ */
+function computeSubSection(pose, currentLat, priorLat, currentLng, priorLng) {
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:523]: I think `currentLat, currentLng, priorLat, priorLng` would be a better order. We're conditioned to see lat,lng pairs together.

File: src/main/webapp/2Dscript.js [hunk 24]
```diff
@@ -80,7 +175,418 @@ function addCenterControl(controlDiv) {
...
+  // Lat can be from [-90,90].
+  let discoveredMinLat = 91;
+  let discoveredMinLatPair = 181;
+  // Lng can be from [-180,180].
+  let discoveredMaxLng = -181;
+  let discoveredMaxLngPair = -91;
+  subLine = [];
+  for (point of pose) {
+    const loopLat = point.lat;
+    const loopLng = point.lng;
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:538]: I don't think these variable carry their weight. Just refer to `point.lat` and `point.lng` within the loop, so the reader doesn't have to keep another two variables in mind.

File: src/main/webapp/2Dscript.js [hunk 25]
```diff
@@ -80,7 +175,418 @@ function addCenterControl(controlDiv) {
...
+ * @param {number} valLat
+ * @param {number} valLng
+ * @return {boolean}
+ */
+function withinBound(minLat, maxLat, minLng, maxLng, valLat, valLng) {
+  if (valLat >= minLat && valLat <= maxLat &&
+    valLng >= minLng && valLng <= maxLng) {
+    return true;
+  }
+  return false;
```
Comment (jmchristen) [inline src/main/webapp/2Dscript.js:591]: This can be simplified:
`return lat >= minLat && lat <= maxLat && lng >= minLng && lng <= maxLng;`

File: src/main/webapp/2Dscript.js (hunk 20 above)
Comment (D-moe) [reply src/main/webapp/2Dscript.js:343]: This was a mistake in the type annotation, where dataEntries is of type PoseData not Point and this loop removes those extra parameters.

Comment (D-moe) [reply src/main/webapp/2Dscript.js:343]: I fixed this error in the type annotation

File: src/main/webapp/2Dscript.js [hunk 26]
```diff
@@ -1,76 +1,171 @@
...
+
+  initialPoseMap = getPolyline(formatPoseData(initialPose), '#FF0000', 1.0, 2);
+  initialPoseMap.setMap(map);
+
+  const centerControlDiv = centerControl();
   map.controls[google.maps.ControlPosition.TOP_CENTER].push(centerControlDiv);
+
+  // Add event listeners to record latitude and longitude globally for selection
+  // box.
+  map.addListener('mousemove', function(event) {
```
Comment (freliciatucker) [inline src/main/webapp/2Dscript.js:107]: just a question: is there any benefits for mousemove over click events?

File: src/main/webapp/2Dscript.js [hunk 27]
```diff
@@ -1,76 +1,171 @@
...
+  // function called somehow, according to stack trace.
+  const initialPose = runs[id].data;
+  const initialPoseMap = runs[id].map;
+  if (initialPose === undefined) {
+    return;
+  }
+
   map = new google.maps.Map(document.getElementById('map'), {
-    center: {lat: pose[0].lat, lng: pose[0].lng},
+    center: {lat: initialPose[0].lat, lng: initialPose[0].lng},
```
Comment (freliciatucker) [inline src/main/webapp/2Dscript.js:95]: could also display timestamp with initalPose[i].gpsTimeStamp

File: src/main/java/proto/SensorData.java [hunk 28]
```diff
@@ -5,20 +5,21 @@
...
   private SensorData() {}
-  public static void registerAllExtensions(
-      com.google.protobuf.ExtensionRegistryLite registry) {
-  }
 
-  public static void registerAllExtensions(
-      com.google.protobuf.ExtensionRegistry registry) {
-    registerAllExtensions(
-        (com.google.protobuf.ExtensionRegistryLite) registry);
+  public static void registerAllExtensions(com.google.protobuf.ExtensionRegistryLite registry) {}
```
Comment (jmchristen) [inline src/main/java/proto/SensorData.java:9]: This file's changes seem unrelated, can you split into another PR?

Comment (D-moe) [reply src/main/java/proto/SensorData.java:9]: This is the result of mvn spotless:apply which Jie mentioned to pass the Java check. This issue arose because some of the code in master (apparently these lines) wasn't in the appropriate format. I think it's fine for everyone to run mvn spotless:apply and commit even if it changes this file multiple times as git will recognize the change as the same and we should likely diligently run the command anyways and that way all checks will pass so we can prevent accidentally putting code in master that is in the improper format.
